org.osgi.util.pushstream/src/org/osgi/util/pushstream/ThresholdPushbackPolicy.java

```diff
@@ -0,0 +1,133 @@
+	private final long	increment;
+
+	/**
+	 * A simple configuration with an initial back pressure of one and and
+	 * increase increment size of one.
+	 * 
+	 * @see #createThresholdPushbackPolicy(int threshold, long initial, long
+	 *      increment)
+	 * @param threshold the queue size limit after which back pressure will be
+	 *            applied
+	 * @throws IllegalArgumentException if any of the given values is lower then
+	 *             zero
+	 * @return a new {@link ThresholdPushbackPolicy}
+	 */
+	public static <T, U extends BlockingQueue<PushEvent< ? extends T>>> ThresholdPushbackPolicy<T,U> createSimpleThresholdPushbackPolicy(
+			int threshold) {
+		return new ThresholdPushbackPolicy<>(threshold, 1L, 1L);
+	}
+
+	/**
+	 * A simple configuration, where the increment size is used as the initial
+	 * back pressure.
+	 * 
+	 * @see #createThresholdPushbackPolicy(int threshold, long initial, long
+	 *      increment)
+	 * @param increment the increments in which the pressure increases
+	 * @param threshold the queue size limit after which back pressure will be
+	 *            applied
+	 * @throws IllegalArgumentException if any of the given values is lower then
+	 *             zero
+	 * @return a new {@link ThresholdPushbackPolicy}
+	 */
+	public static <T, U extends BlockingQueue<PushEvent< ? extends T>>> ThresholdPushbackPolicy<T,U> createIncrementalThresholdPushbackPolicy(
+			int threshold, long increment) {
+		return new ThresholdPushbackPolicy<>(threshold, increment, increment);
+	}
+
+	/**
+	 * @param increment the increments in which the pressure increases.
+	 * @param initial the initial back pressure which defines the floor after
+	 *            the threshold is exceeded
+	 * @param threshold the queue size limit after which back pressure will be
+	 *            applied
+	 * @throws IllegalArgumentException if any of the given values is lower then
+	 *             zero
+	 * @return a new {@link ThresholdPushbackPolicy}
+	 */
+	public static <T, U extends BlockingQueue<PushEvent< ? extends T>>> ThresholdPushbackPolicy<T,U> createThresholdPushbackPolicy(
+			int threshold, long initial, long increment) {
+		return new ThresholdPushbackPolicy<>(threshold, initial, increment);
+	}
+	
+	/**
+	 * @param increment the increments in which the pressure increases.
+	 * @param initial the initial back pressure which defines the floor after
+	 *            the threshold is exceeded
+	 * @param threshold the queue size limit after which back pressure will be
+	 *            applied
+	 * @throws IllegalArgumentException if any of the given values is lower then
+	 *             zero
+	 */
+	private ThresholdPushbackPolicy(int threshold, long initial,
+			long increment) {
+		if (threshold < 0 || initial < 0L || increment < 0L) {
+			throw new IllegalArgumentException(String.format(
+					"threshold, initial and increment must each be larger than zero. Current values are threshold [%s], initial [%s], increment [%s]",
```

Comment:
The code requires non-negative values, but the exception messages says you must use positive values. This needs to be reconciled. My guess is you really want positive values, so the code test must be `<= 0'`.

The new TCK tests should also confirm that 0 is illegal to match this code.

Reply:
0 for any or all of its values is actually legal and having one or more of them as such might even make sense in some situation. Thus I allowed for it. One might argue that e.g. all initialized with 0 does not make sense, as it would effectively do nothing, but it is still legal and work fine.

Reply:
> The new TCK tests should also confirm that 0 is illegal to match this code.

This should not be the case, at least for `initial` and `increment`. The JavaDoc for this class outlines uses `0` for both of these (not simultaneously) to get useful behaviours. Specifically a zero increment gives a “threshold fixed” policy, and a zero initial gives a “threshold linear” policy. 

As Jürgen points out, zero for both initial and increment is a bit strange but not actively broken. Zero for threshold collapses to a combination of linear plus fixed, and might be considered odd enough to be broken.

Reply:
OK, then the exception message is wrong since it states the values "must each be larger than zero".

Reply:
That is correct. Changed it accordingly.